### PR TITLE
Adding the direct_message model

### LINKED ISSUE #22: Adding the message model
Adding the message model including:
- creating the message table (as a model)
- testing the message model

### PR DESCRIPTION
**Pull request by TomerNewmanPrograms**
- Adding direct_message model
- Adding validation tests for it.
fix #22

### CODE REVIEW

**Review by OmerCS8 — DISMISSED** [review]
LGTM

**Review by MatanP12 — COMMENTED** [review]
I think we should delete a message only after both the sender and receiver users were deleted.

**Comment by TomerNewmanPrograms** [comment]
> I think we should delete a message only after both the sender and receiver users were deleted.

Why do we need to save the message if only one of them is deleted?

**Comment by OmerCS8** [comment]
> I think we should delete a message only after both the sender and receiver users were deleted. 

I belive that the way it's right now is more consistent with the rest of our project (our decision that when a user is being deleted we also delete his posts and comments).

**Comment by MatanP12** [comment]
> > I think we should delete a message only after both the sender and receiver users were deleted.
> 
> Why do we need to save the message if only one of them is deleted?

You are right I think it will make our work more complicated.

**Review by MatanP12 — APPROVED** [review]
LGTM

File: direct_message/tests.py [hunk 1]
```diff
@@ -1,3 +1,82 @@
...
+
+@pytest.fixture
+def message0(new_user_sender, new_user_reciever):
+    return models.Message(sender=new_user_sender,
+                          reciever=new_user_reciever,
+                          creation_date=CREATION_DATE,
+                          message_text='test')
+
+
+def save_a_message(message0):
```
[the remark this answers is not in the document]
Comment (TomerNewmanPrograms) [reply direct_message/tests.py:30]: Instead of saving the receiver, sender and then the message over and over again I wrote a function for it

**Review by natifridman — COMMENTED** [review]
Is this part of the ERD? I don't see it in the ERD and DFD

**Comment by TomerNewmanPrograms** [comment]
> Is this part of the ERD? I don't see it in the ERD and DFD

Because we chose to focus before this weekend on the Post and Resume, the direct_message isn't a part of the ERD and DFD.
We are changing it now

File: direct_message/tests.py (hunk 1 above)
Comment (yulisuliman) [inline direct_message/tests.py:30]: I think you should add  `@pytest.mark.django_db()` annotation above this function.

Comment (TomerNewmanPrograms) [reply direct_message/tests.py:30]: The save_a_message function is being called inside other functions that use `@pytest.mark.django_db()`

File: direct_message/tests.py [hunk 2]
```diff
@@ -1,3 +1,82 @@
...
+
+
+def is_empty_string(string):
+    for word in string:
+        if word != ' ':
+            return False
+    return True
+
+
+@pytest.mark.parametrize("empty_string", ['', ' ', "     "])
```
[the remark this answers is not in the document]
Comment (TomerNewmanPrograms) [reply direct_message/tests.py]: fixed

File: setup.sh [hunk 3]
```diff
@@ -12,5 +12,5 @@ pipenv sync --dev
 
 # run our app. setsit, the parentheses and "&" are used to perform a "double
 # fork" so that our app stays up after the setup script finishes.
-# The app logs are redirected to the `runserver.log` file.
+# The app logs are redirected to the 'runserver.log' file.
```
[the remark this answers is not in the document]
Comment (TomerNewmanPrograms) [reply setup.sh]: No, I reverted the change

File: direct_message/tests.py [hunk 4]
```diff
@@ -1,3 +1,82 @@
...
+                          message_text='test')
+
+
+def save_a_message(message0):
+    message0.sender.save()
+    message0.reciever.save()
+    message0.save()
+
+
+def test_new_message_fields(message0, new_user_sender, new_user_reciever):
```
[the remark this answers is not in the document]
Comment (TomerNewmanPrograms) [reply direct_message/tests.py]: I changed it to `test_message_constructor`

File: direct_message/tests.py [hunk 5]
```diff
@@ -1,3 +1,82 @@
...
+    return True
+
+
+@pytest.mark.parametrize("empty_string", ['', ' ', '     '])
+def test_is_empty_message(new_message, empty_string):
+    new_message.message_text = empty_string
+    assert is_empty_string(new_message.message_text)
+
+
+@pytest.mark.django_db()
```
[the remark this answers is not in the document]
Comment (TomerNewmanPrograms) [reply direct_message/tests.py:56]: I thought about it, the problem is that after each save (and assertion) there is an need for each function to delete the new  instance.

**Review by MatanP12 — DISMISSED** [review]
LGTM

**Review by yulisuliman — APPROVED** [review]
Good job!

File: direct_message/models.py [hunk 6]
```diff
@@ -1,3 +1,13 @@
-# from django.db import models
+from django.db import models
+from django.utils import timezone
+from django.contrib.auth.models import User
 
-# Create your models here.
+
+class Message(models.Model):
+    sender = models.ForeignKey(User, on_delete=models.CASCADE, related_name='sender')
+    reciever = models.ForeignKey(User, on_delete=models.CASCADE, related_name='reciever')
```
Comment (natifridman) [inline direct_message/models.py:8]: receiver

File: direct_message/tests.py [hunk 7]
```diff
@@ -1,3 +1,82 @@
...
+
+
+@pytest.fixture
+def new_user_sender():
+    return User(username='TestSender', first_name='Nick',
+                last_name='Birch', password='1234', email='Nick@email.com')
+
+
+@pytest.fixture
+def new_user_reciever():
```
Comment (natifridman) [inline direct_message/tests.py:17]: receiver

File: direct_message/tests.py [hunk 8]
```diff
@@ -1,3 +1,82 @@
...
+    return User(username='TestReciever', first_name='Andrew',
```
Comment (natifridman) [inline direct_message/tests.py:18]: receiver

**Review by natifridman — CHANGES_REQUESTED** [review]
Please fix the typo for `receiver`

File: direct_message/migrations/0001_initial.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='Message',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('creation_date', models.DateTimeField(default=django.utils.timezone.now)),
+                ('message_text', models.TextField()),
+                ('reciever', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE,
```
[the remark this answers is not in the document]
Comment (TomerNewmanPrograms) [reply direct_message/migrations/0001_initial.py]: fixed

File: direct_message/tests.py (hunk 8 above)
Comment (TomerNewmanPrograms) [reply direct_message/tests.py]: fixed

File: direct_message/models.py (hunk 6 above)
Comment (TomerNewmanPrograms) [reply direct_message/models.py]: fixed

**Review by MatanP12 — APPROVED** [review]
LGTM

**Review by yulisuliman — APPROVED** [review]

File: direct_message/tests.py [hunk 10]
```diff
@@ -1,3 +1,82 @@
...
+                          message_text='test')
+
+
+def save_a_message(new_message):
+    new_message.sender.save()
+    new_message.receiver.save()
+    new_message.save()
+
+
+def test_message_constructor(new_message, new_user_sender, new_user_receiver):
```
Comment (natifridman) [inline direct_message/tests.py:36]: Maybe we can change this test name to be more clear? if this checks a new message maybe we can call something like `test_new_message`?

Comment (TomerNewmanPrograms) [reply direct_message/tests.py]: fixed to 'new_message'

**Review by MatanP12 — APPROVED** [review]
LGTM

**Review by yulisuliman — DISMISSED** [review]

**Review by AlonShakaroff — DISMISSED** [review]
LGTM

File: direct_message/tests.py (hunk 10 above)
Comment (natifridman) [reply direct_message/tests.py]: I tend to agree with Yuli

Comment (TomerNewmanPrograms) [reply direct_message/tests.py]: Changed, I also changed a couple of more names to be more consistent with Yuli's suggestion @natifridman  @yulisuliman

**Review by AlonShakaroff — APPROVED** [review]

**Review by yulisuliman — APPROVED** [review]

**Review by OmerCS8 — APPROVED** [review]
LGTM

**Review by natifridman — APPROVED** [review]

**Pull request merged**